PYIC-7076: Add check-reverification-identity lambda

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Add check-reverification-identity lambda

### Why did it change

We need to ensure that if a user starts a reverification journey that they have some form of existing identity to re-verify. This adds a new lamdba that will check if they've got a P2, P1 or operational identity. It's not concerned about CIs, expired VCs, account interventions or pending identities.


### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-7076](https://govukverify.atlassian.net/browse/PYIC-7076)


[PYIC-7076]: https://govukverify.atlassian.net/browse/PYIC-7076?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-7076: Add check-reverification-identity lambda

We need to ensure that if a user starts a reverification journey that
they have some form of existing identity to re-verify. This adds a new
lamdba that will check if they've got a P2, P1 or operational identity.
It's not concerned about CIs, expired VCs, account interventions or
pending identities.

Joe-Edwards-GDS commented on lambdas/check-reverification-identity/src/main/java/uk/gov/di/ipv/core/checkreverificationidentity/CheckReverificationIdentityHandler.java:
```diff
@@ -0,0 +1,197 @@
+        }
+    }
+
+    private boolean vcsContainIdentity(List<VerifiableCredential> vcs)
```
This seems very similar to `CheckExistingIdentityHandler#getStrongestAttainedVotForVtr` (with all valid VOTs and no CIs) - do you think there's any merit in trying to commonise the two? Either by merging the lambdas or pulling out a shared service...

There might be a question about whether we _need_ to check for a profile here at all - in principle we should only have an identity in EVCS if they match(ed) a valid profile. It probably doesn't hurt - and should be fairly speedy (if we don't call CIMIT), so I'm not wildly fussed.

Wynndow replied:
They seem similar because I used it to inform this code! I figured that they were different enough though to stand on their own. Can look at pulling out a common service but I suspect it may be more trouble than it's worth.

I'd rather not combine the two lambdas - check-existing-identity is already pretty complicated and sprawling, and shoehorning another code path through it gives me the fear.

Wynndow committed:
PYIC-7076: Refactor vot matching to separate class

This moves the logic for matching a VOT from a given set of VC to a
shared class that is now used by the check-existing-identity handler as
well as the check-reverification-identity class.

Wynndow committed:
PYIC-7076: Set target VOT in check-reverification-identity

Rather than hardcoding reverification journeys to P2, we can determine
it from the identity to be reverified.

Wynndow commented on lambdas/check-reverification-identity/src/main/java/uk/gov/di/ipv/core/checkreverificationidentity/CheckReverificationIdentityHandler.java:
```diff
@@ -0,0 +1,186 @@
+            }
+
+            if (GPG45.equals(reverificationIdentityVot.get().vot().getProfileType())) {
+                ipvSession.setTargetVot(reverificationIdentityVot.get().vot());
```
Do we actually care about this? If the previous identity was a P2, do we _need_ the user to prove their identity to a P2 level? Would P1 give us enough confidence to let them use their MFA again?

Joe-Edwards-GDS replied:
IMO this is kinda meaningless - the user doesn't prove to any GPG45 level (e.g. there's no fraud score!)

The only places it's used is to determine the CI threshold, and if we include F2F or KBV, then the scores being requested.

I wonder if we should approach this in a different way - either setting it to P2 automatically, or deliberately leaving it null and handling explicitly in all of those places.

(cc @DanCorderIPV who I think spotted this issue originally)

Wynndow replied:
Looking at the reverification journey, we'll never touch any of those things (checking if breaching CI threshold, or determining evidence required for F2F/KBV). So my inclination is to just leave it null and then handle.

Joe-Edwards-GDS replied:
If you're confident that we'll never hit it, then that might be cleaner.

It does feel like we'd be leaving a little bit of a foot-gun if we change it later (e.g. to introduce KBVs), but I suppose we would hit it pretty quickly.

The remaining use-case that I'm not sure about is for the TICF CRI (i.e. `IpvSessionItem#getThresholdVot`)

DanCorderIPV replied:
That sounds reasonable to me - I suspect we just set it to P2 originally because things broke if we didn't. If you're going to check those areas and make them not break then that should be OK.

Wynndow replied:
I'm gonna row back on my idea of leaving it as null. I failed to realise that we'll be checking for CIs as part of the return from a CRI, as well as in TICF, which requires a VOT.
We'll either need a mechanism to skip the CI check (which doesn't feel right), or decide on a threshold.

I don't want to put too much info about scores here, so I think lets talk about this offline.

Joe-Edwards-GDS commented on lambdas/check-reverification-identity/src/main/java/uk/gov/di/ipv/core/checkreverificationidentity/CheckReverificationIdentityHandler.java:
```diff
@@ -0,0 +1,186 @@
+        var gpg45Vcs = VcHelper.filterVCBasedOnProfileType(vcs, GPG45);
+        var gpg45Scores = gpg45ProfileEvaluator.buildScore(gpg45Vcs);
+        var gpg45VcsCorrelated = userIdentityService.areVcsCorrelated(gpg45Vcs);
+        var operationalVcs = VcHelper.filterVCBasedOnProfileType(vcs, OPERATIONAL_HMRC);
```
Should we do the filtering in the `VotMatcher` - looks quite similar for both usages?

Wynndow replied:
I intentionally had them being passed in as we use the filtered VCs and the gpg45 scores elsewhere in the check-existing-identity lambda (for sending an audit event). I didn't want to have to do the filtering again, or send an audit event with details that weren't acutally used when checking the VOT.
Another option is we send that audit event (`IPV_GPG45_PROFILE_MATCHED`) in the new lambda but I figured it's not required and may be confusing.

Joe-Edwards-GDS replied:
Ok sure - I think the overhead of filtering the list twice is pretty negligible, so I'd go for whichever looks neatest. The GPG45 scores are a bit more awkward (still pretty cheap to calculate, but worse than the filtering) :(

In terms of the audit event, I'm not sure! I could probably justify either way... If you want to retain the control in the calling code (which makes sense), I suppose you could have the method return a triple of Vot/GPG45Profile/GPG45Scores or something. Perhaps call it something like `ProfileMatchingResult` if it makes you feel better 😄

Joe-Edwards-GDS replied:
Overall I quite like `ProfileMatchingResult matchProfile(List<Vot> vots, List<VerifiableCredential> vcs, List<ContraIndicator> cis)` as a good clean signature!

Wynndow replied:
It wasn't so much about the overhead of filtering, more that I wanted to ensure that the _same_ list used to match the VOT was used in the audit event. If we recalculate it for the audit event if feels like there's an opportunity in future for those lists to get differentiated somehow. I'm probably being over cautious. I'll do the triple thing then.

Joe-Edwards-GDS commented on lambdas/check-reverification-identity/src/main/java/uk/gov/di/ipv/core/checkreverificationidentity/CheckReverificationIdentityHandler.java:
```diff
@@ -0,0 +1,186 @@
+                return NOT_FOUND_RESPONSE;
+            }
+
+            if (GPG45.equals(reverificationIdentityVot.get().vot().getProfileType())) {
```
(regardless of what we do below, I think we still need to do something with the operation profiles!)

Wynndow replied:
if we do nothing (leave taget VOT as null), what else is there to do?

Joe-Edwards-GDS commented on libs/user-identity-service/src/main/java/uk/gov/di/ipv/core/library/service/VotMatcher.java:
```diff
@@ -0,0 +1,111 @@
+                    return Optional.of(new VotAndProfile(vot, matchedGpg45Profile.get()));
+                }
+            }
+            if (hasOperationalProfileVc(vot, operationalVcs, contraIndicators)) {
```
Should this be an else, so we don't try and check this for GPG45 VOTs?

Joe-Edwards-GDS reviewed (commented):
I think this feels better, and probably more reusable - what do you think?

(And takes some of the logic out of CheckExistingIdentity, which only feels like a good thing)

Wynndow committed:
PYIC-7076: Don't set targetVot for reverification journeys

It's not used in the journey as we don't actually do a gpg45 check, and
it makes no sense if the user has an operational profile.

This means that it's now possible for that value to be null in a users
session, so we handle that possibility where we read it.

Wynndow committed:
PYIC-7076: Refactor Vot matcher

Clean up the signature of the method a bit.

Wynndow committed:
PYIC-7076: Skip CI check with TICF

To keep inline with how we handle CIs when a user is on a reverification
journey.

Joe-Edwards-GDS reviewed (approved):
Thanks, I think this is a bit more consistent now
